### PR TITLE
Issue #261: adding missing dense descriptor and delete unnecessary se…

### LINKED ISSUE #261: PageRank algorithm optimisations
The PageRank implementation does not pass the dense descriptor when it could, and includes the use of the expensive `invert_mask` in its performance-critical loop.

### PR DESCRIPTION
**Pull request by aristeidis-mastoras**
Closes issue #261 .

Closes issue #305 .

* [x] Adding `dense` descriptor
* [x] Spreading the `descriptor` template parameter in the `eWiseApply` and `eWiseLambda` functions 
* [x] work around invert_mask

### CODE REVIEW

File: include/graphblas/base/blas1.hpp [hunk 1]
```diff
@@ -3737,15 +3737,15 @@ namespace grb {
...
 		typename Func,
 		typename DataType,
 		Backend backend,
-		typename Coords,
-		typename... Args
+		typename Coords
 	>
 	RC eWiseLambda(
 		const Func f,
-		const Vector< DataType, backend, Coords > & x, Args...
```
Comment (anyzelman) [inline include/graphblas/base/blas1.hpp:3748]: Why are the varargs removed here?

File: include/graphblas/hyperdags/blas2.hpp [hunk 2]
```diff
@@ -565,9 +569,13 @@ namespace grb {
 			return vxm< descr >( u, mask, v, A, ring, phase );
 		}
 		const RC ret = vxm< descr >(
-			internal::getVector(u), internal::getVector(mask),
-			internal::getVector(v), internal::getVector(v_mask), internal::getMatrix(A),
-			ring, phase
+			internal::getVector(u),
+			internal::getVector(mask),
```
Comment (anyzelman) [inline include/graphblas/hyperdags/blas2.hpp]: I think it's clearer if these two are grouped (and similar further down for u and v_mask--- I agree the matrix could/should go on its own row)

File: include/graphblas/nonblocking/blas2.hpp [hunk 3]
```diff
@@ -1518,11 +1519,12 @@ namespace grb {
 		internal::le.execution();
 
 		// second, delegate to the reference backend
-		return eWiseLambda< ActiveDistribution, Func, DataType, RIT, CIT, NIT >(
+		return eWiseLambda< descr, ActiveDistribution, Func, DataType, RIT, CIT, NIT >(
 			f, internal::getRefMatrix( A ), s, P );
 	}
 
 	template<
+		Descriptor descr = descriptors::no_operation,
```
Comment (anyzelman) [inline include/graphblas/nonblocking/blas2.hpp:1523]: This descriptor is not passed to the subsequent call to eWuseLambda (line 1551 below)

File: include/graphblas/reference/blas1.hpp [hunk 4]
```diff
@@ -9759,7 +9760,10 @@ namespace grb {
 	 * @see Vector::operator[]()
 	 * @see Vector::lambda_reference
 	 */
-	template< typename Func, typename DataType, typename Coords >
+	template<
+		Descriptor descr = descriptors::no_operation,
```
Comment (anyzelman) [inline include/graphblas/reference/blas1.hpp]: The function should return illegal if x was not dense but the dense descriptor was passed (maybe this check is also missing with the other backends)

File: include/graphblas/reference/blas2.hpp [hunk 5]
```diff
@@ -2824,6 +2825,7 @@ namespace grb {
 	 * @see grb::eWiseLambda for the user-level specification.
 	 */
 	template<
+		Descriptor descr = descriptors::no_operation,
```
Comment (anyzelman) [inline include/graphblas/reference/blas2.hpp:2854]: (it seems forgotten here as well -- note that the dense descriptor is only defined to apply for vectors (x), and not for A)

**Review by anyzelman — CHANGES_REQUESTED** [review]
Please see the review comments -- also one question: should allowing a dense descriptor to eWiseLambda not also lead to a change in a cpp file src/nonblocking/ somewhere?

**Comment by anyzelman** [comment]
Note: I also added the todo of removing the invert_mask in the PR algorithm. If no objection I could do that quickly around now?

**Comment by anyzelman** [comment]
> Note: I also added the todo of removing the invert_mask in the PR algorithm. If no objection I could do that quickly around now?

This is done. I did run into a new bug in nonblocking, I opened issue #305 for it.

File: include/graphblas/base/blas1.hpp (hunk 1 above)
Comment (byjtew) [reply include/graphblas/base/blas1.hpp:3748]: Because they are never implemented

**Comment by anyzelman** [comment]
@aristeidis-mastoras @byjtew -- this one is now conflicted but was the next on my list. I'll try resolve the conflicts now

**Comment by anyzelman** [comment]
Rebase complete, smoke tests with LPF OK. Unit tests with LPF and GitHub CI running overnight

**Comment by anyzelman** [comment]
(If anyone of @aristeidis-mastoras @byjtew wants to review how I handled the conflicts lemme know before tomorrow -- if no message by e.g. noon, I'll proceed with review/merge:)

**Comment by anyzelman** [comment]
Crossref #305 which this MR should also fix (TBC)

**Review by anyzelman — DISMISSED** [review]

**Comment by anyzelman** [comment]
Review OK but found one issue in the bsp1d backend. Re-running smoke & unit tests with LPF.

**Comment by anyzelman** [comment]
All tests OK, will merge

**Comment by anyzelman** [comment]
Concept release notes:

Prior to this MR, the PageRank implementation does not pass the dense descriptor when it could, while it includes the use of the expensive `invert_mask` in its performance-critical loop. Mask inversion is expensive since it requires iterating over the full vector range, while, in this case, the inverted mask is constant. Hence, this MR introduces the following changes:
- the inverted mask is precomputed once, while in the performance-critical loop the precomputed inversion is reused as a structural mask; and
- the PR algorithm is reviewed to pass the dense descriptor whenever possible.

This improvement does come at the cost of one additional work-space vector that is now required for the PageRank implementation.

In implementing these fixes, the following feature has been added:
- the `grb::eWiseLambda` now takes a descriptor template argument. This was previously accepted in the nonblocking backend, but was missing from the specification as well as missing in the implementations of other backends.

The following bugs were encountered and fixed with this MR:
- mixed-domain vector-to-scalar reductions were bugged for the nonblocking backend: when the second (underlying) monoid domain type did not match its third (output) domain type, a compiler error would be returned. This MR fixes the behaviour for both left- and right-handed reductions;
- the descriptors for multiple-vector eWiseLambdas were not correctly propagated for the nonblocking backend.

**Review by anyzelman — APPROVED** [review]

**Comment by anyzelman** [comment]
Found one more minor issue in nonblocking, now fixed -- re-running unit tests excluding the LPF ones.

**Pull request merged**